### ISSUE TITLE
Training details for models involving Non-local blocks

### ISSUE THREAD

**Issue by xyang35** [facebookresearch/SlowFast#57]
Hi, thanks for your great work! I have a question about the training details for models with NL blocks. In the paper, you mentioned "For all models involving NL, we initialize them with the counterparts that are trained without NL, to facilitate convergence." Could you elaborate more on this part? For example, the initial learning rate, the number of epochs, whether the output layer (FC layer) also initialized with the counterparts?  Thank you in advance for your help!

**Comment by takatosp1** [comment]
Hi, thanks for using the codebase!
I guess you are referring to the non-local paper? (would be good if you can provide more context since this codebase includes multiple papers :P )

If I understand you correctly, basically what does it meant is we train the backbone without nonlocal first, then add the nonlocal and retrain (or if you prefer to call it finetune).

I will release the nln-local pretrain yaml in the next release in 1 week.

**Comment by xyang35** [comment]
Thanks for your reply, but I still cannot find the answer from the latest release. The yaml files of the models with or without NL are almost identical except the configurations of non-local blocks.

The training details I mentioned before referred to the appendix of the slowfast paper. I'm just wondering what the training schedule is for the retraining part. Is the retraining (with NL) using the same learning rate schedule and the same number of epochs as those for training the backbone without nonlocal?

**Comment by takatosp1** [comment]
Hi,
thanks for the follow up! The NLN recipe has been released long time ago. You should able to find the config files at
`C2D_NLN_8x8_R50_IN1K.yaml` `I3D_NLN_8x8_R50_IN1K.yaml` which you can use to reproduce the non-local with imagenet pretrain inflation. Remember to use the ImageNet pretrain I shared [here](https://github.com/facebookresearch/SlowFast/blob/master/MODEL_ZOO.md).

**Comment by takatosp1** [comment]
If you need some settings other than what I shared above, could you share me a more clear reference?
For example, which result from the non-local paper you want to reproduce?

For example, I borrowed the number sheet from the [nonlocal repo](https://github.com/facebookresearch/video-nonlocal-net):
script | input frames | freeze bn? | 3D conv? | non-local? | top1 | in paper | top5 | model | logs
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
run_c2d_baseline_400k_32f.sh | 32 | - | - | - | 72.0 | 71.8 | 90.0 | link | link
run_c2d_nlnet_400k_32f.sh | 32 | - | - | Yes | 73.9 | 73.8 | 91.0 | link | link
run_i3d_baseline_400k_32f.sh | 32 | - | Yes | - | 73.6 | 73.3 | 90.8 | link | link
run_i3d_nlnet_400k_32f.sh | 32 | - | Yes | Yes | 74.9 | 74.9 | 91.6 | link | link
run_i3d_baseline_affine_400k_128f.sh | 128 | Yes | Yes | - | 75.2 | 74.9 | 92.0 | link | link
run_i3d_nlnet_affine_400k_128f.sh | 128 | Yes | Yes | Yes | 76.5 | 76.5 | 92.7 | link | link


input frames | freeze bn? | 3D conv? | non-local? | top1 | in paper | top5 | model | logs
-- | -- | -- | -- | -- | -- | -- | -- | --
run_c2d_baseline_400k_32f.sh | 32 | - | - | - | 72.0 | 71.8 | 90.0 | link | link
run_c2d_nlnet_400k_32f.sh | 32 | - | - | Yes | 73.9 | 73.8 | 91.0 | link | link
run_i3d_baseline_400k_32f.sh | 32 | - | Yes | - | 73.6 | 73.3 | 90.8 | link | link
run_i3d_nlnet_400k_32f.sh | 32 | - | Yes | Yes | 74.9 | 74.9 | 91.6 | link | link
run_i3d_baseline_affine_400k_128f.sh | 128 | Yes | Yes | - | 75.2 | 74.9 | 92.0 | link | link
run_i3d_nlnet_affine_400k_128f.sh | 128 | Yes | Yes | Yes | 76.5 | 76.5 | 92.7 | link | link

**Comment by xyang35** [comment]
I'm looking for settings similar to "SlowFast 16×8, R101+NL" (the last row in Table 2. of the **SlowFast** paper). I think the model was not trained from ImageNet pretrained weights, and now I'm a little confused whether the result is obtained by 
1) fully trained from scratch using the config `SLOWFAST_NLN_16x8_R101_50_50.yaml` or, 
2) first initialized with the counterpart without NL, then retrained with NL, as mentioned in the first paragraph of the appendix of SlowFast paper.

If the answer is (2), then I'm thinking maybe the training schedule of the retraining part with NL would be different from that of the training w/o NL, but I cannot find that in the current codebase.

Thanks again for your help.

**Comment by nik13** [comment]
@xyang35, were you able to train it? Any pretrained weights available?
